highfidelity/hifi
# Shader load time performance optimiziations

* Cache shader binaries
* Optimize loading shader uniforms
* Add unit test for benchmarking shader load performance
* Fixup unit tests to use test-utils library

## Testing

This PR will cache the shaders we use as binary data after the first run.  The cache will be located in the application data directory, in the case of this PR, on Windows that would be `~/AppData/High Fidelity - PR13082/Interface/shaders/cache.json`.  

### Windows testing

Testing will require both an nVidia and an AMD based machine

* On the nVidia machine, install the PR build.  Verify that the above file does not exist.
* Run Interface
  * Verify the file still does not exist
  * Verify rendering is normal
  * Switch between desktop and HMD or stereo display modes.  Verify rendering is normal
* Exit Interface
* Verify the shader cache file exists and is populated (it should be about 10MB on a windows machine with recent nVidia drivers)
* Restart interface.  
  * The application should move from the grey screen to a rendered scene faster than before (although the difference might be very slight)
  * Re-follow the same steps as before for validating rendering
* Exit interface
* Copy the shader cache from the above location to a flash drive.
* On the AMD machine, install the PR
* Verify the shader cache does not exist
* Copy the shader cache from the flash drive to the AMD machine
* Run interface (on the AMD system)
  * Verify that interface starts normally.  
  * Run the same rendering validation steps for desktop and HMD/Stereo
* Exit interface
* Restart interface (on the AMD system)
  * The application should move from the grey screen to a rendered scene faster than before (although the difference might be very slight)
  * Run the same rendering validation steps for desktop and HMD/Stereo
* Exit interface

### Mac testing

Install and test run the PR build.  The Mac does not support cached binary shaders for OpenGL, so the file may or may not be created, but will likely be an empty JSON object if it exits.

* Verify that the application starts up and shuts down successfully several times
* The application startup time should be significantly reduced compared to the current master build.  

### Unit test tool

https://hifi-public.s3.amazonaws.com/austin/ShaderLoadTest.zip
